internal/app/machined/pkg/controllers/network/operator/dhcp4.go

```diff
@@ -61,6 +65,49 @@ func (d *DHCP4) Prefix() string {
 	return fmt.Sprintf("dhcp4/%s", d.linkName)
 }
 
+// hostnameStatusMetadata represents the full metadata of any version of a network.HostnameStatus.
+var hostnameStatusMetadata = resource.NewMetadata(network.NamespaceName, network.HostnameStatusType, network.HostnameID, resource.VersionUndefined)
+
+// setupHostnameWatch returns a channel that outputs all events related to hostname changes.
+func (d *DHCP4) setupHostnameWatch(ctx context.Context) (<-chan state.Event, error) {
+	hostnameWatchCh := make(chan state.Event)
+
+	return hostnameWatchCh, d.state.WatchKind(ctx, hostnameStatusMetadata, hostnameWatchCh)
+}
+
+// updateHostname update the given pointer with the current node hostname.
+func (d *DHCP4) updateHostname(ctx context.Context, hostname **string) error {
+	hostnameResource, err := d.state.Get(ctx, hostnameStatusMetadata)
+	if err != nil {
+		if state.IsNotFoundError(err) {
+			*hostname = nil
+
+			return nil
+		}
+
+		return err
+	}
+
+	*hostname = &hostnameResource.(*network.HostnameStatus).TypedSpec().Hostname
+
+	return nil
+}
+
+// knownHostname checks if the given hostname has been defined by this operator.
+func (d *DHCP4) knownHostname(hostname *string) bool {
```

Comment:
This is a good idea. `knownHostname` really is a hack, and it would be useful to know which of the hostname candidates was elected as the node hostname for transparency and debugging as well. It would be best if the `HostnameStatus` source be used to identify an individual `HostnameSpec`, as just knowing the configuration layer of the source is not enough (e.g. ambiguity ensues when both the `dhcp4` and `dhcp6` operators have `HostnameSpec`s defined in the operator layer).

The current functionality with `knownHostname` works as-is if you prefer to merge this PR first, but I can also rebase on top of the changes to `HostnameStatus` if that's better for you.

Reply:
thanks, I think I can get in some changes early next week for that, as we're still waiting for upstream dhcp library fixes?

Reply:
Yes, go for it. I can then rebase on top of that work. There's one more issue with the hostname "parroting" prevention as well (not directly related to `knownHostname`), the approach needs to be changed a bit to work even when the only hostname (apart from the default hostname) is defined by DHCP. I'll look at fixing that early next week.

Reply:
The new approach is now in place and confirmed working even in the new edge cases I found. When the locally defined hostname and the hostname served by DHCP are the same, both now register correctly.